drivers: ieee802154: Rem trx_state variable

nandojve opened the pull request:
The rx timeout timer callback need update **trx_state** variable and this variable is protected by a **mutex**. Because of that, when compiling the system with **CONFIG_ASSERT=y** the system reports _ASSERTION FAIL [!arch_is_in_isr()] @ ZEPHYR_BASE/kernel/include/ksched.h:262_.

This refactor the driver remove **trx_state** variable dependency and consequently removes **phy_mutex** and **rx timeout timer** to be compliant with kernel rules.

Fixes: #23198

This PR includes #22511

nandojve committed:
drivers: ieee802154: rf2xx: Add SRAM read method

Add SRAM read method to enable advanced uses on at86rf2xx driver.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: ieee802154: rf2xx: Add RX improvements

The current version of at86rf2xx RX implementation don't uses advanced
capabilities offer by the transceiver. This access SRAM space to gatter
PHR information in parallel with transceiver frame reception. It allows
improve RX reception by handling properlly the frame protection feature
removing transceiver states changes.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
boards: arm: atsamr21_xpro: Fix corrupted frames on rf2xx

When using internal IEEE 802.15.4 transceiver AT86RF233 at SPI speeds
above 6MHz was detected that frame buffer read returns corrupted data.
This set spi-max-frequency to 6MHz at atsamr21_xpro.dts to ensure best
relation of performance vs reliability.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: ieee802154: rf2xx: Rem trx_state variable

The rx timeout timer callback need update trx_state variable and this
variable is protected by a mutex. Because of that, when compiling the
system with CONFIG_ASSERT=y the system reports 'ASSERTION FAIL
[!arch_is_in_isr()] @ ZEPHYR_BASE/kernel/include/ksched.h:262'.

This refactor the driver remove trx_state variable dependency and
consequently removes phy_mutex and rx timeout timer to be compliant
with kernel rules.

Fixes: #23198

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
The changes were tested with two **atsamr21** boards running `samples/net/socktets/echo_[client|server]`. The system seems to be working without any problems. Below a line printing from server side that shows number of packets after some time.

`[02:19:11.140,000] <inf> net_echo_server_sample: IPv6 UDP: Sent 305000 packets`

@markus-becker-tridonic-com can you check if everything is OK from your side and give us a feedback, please.

nandojve commented:
@jhedberg

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx_iface.c:
```diff
@@ -7,7 +7,11 @@
  */
 
 #define LOG_MODULE_NAME ieee802154_rf2xx_iface
+#if defined(CONFIG_IEEE802154_DRIVER_LOG_LEVEL)
```
Why this ifdef?
#define LOG_LEVEL CONFIG_IEEE802154_DRIVER_LOG_LEVEL  
should be sufficient.

nandojve replied:
Rollback

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx_iface.c:
```diff
@@ -7,7 +7,11 @@
  */
```
in you commit title add "for rf2xx driver"

nandojve replied:
Done.

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -7,7 +7,11 @@
  */
```
add "for rf2x driver" in your commit title

nandojve replied:
Done.

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -7,7 +7,11 @@
  */
 
 #define LOG_MODULE_NAME ieee802154_rf2xx
+#if defined(CONFIG_IEEE802154_DRIVER_LOG_LEVEL)
```
same as previous patch, no need of ifdef.

nandojve replied:
Rollback

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -230,26 +257,35 @@ static void rf2xx_thread_main(void *arg)
+					/* Ensures that automatically ACK will be sent
+					* when requested
+					*/
+					while (rf2xx_iface_reg_read(dev,
```
when you start to have too many if/for/while imbrication, it's time to refactor a bit for readability

nandojve replied:
Done on fd54e41

nandojve committed:
drivers: ieee802154: rf2xx: Refactor rf2xx_thread_main

Current rf2xx_thread_main code have too many if/for/while imbrication.
Extract methods from rx2xx_thread_main for better readability.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

javier-moreno-tridonic-com commented:
@nandojve Hi, I've tested this on our hardware and so far we haven't encountered any more problems.

jukkar reviewed (approved):
LGTM
